src/main/java/de/zalando/aruha/nakadi/controller/EventStreamController.java

```diff
@@ -0,0 +1,170 @@
+                    writeProblemResponse(response, outputStream, HttpStatus.UNPROCESSABLE_ENTITY.value(),
+                            new Problem("stream_timeout can't be lower than batch_flush_timeout"));
+                    return;
+                }
+
+                // deserialize cursors
+                Optional<List<Cursor>> cursors = Optional.empty();
+                if (cursorsStr != null) {
+                    try {
+                        cursors = Optional.of(jsonMapper.<List<Cursor>>readValue(cursorsStr,
+                                new TypeReference<ArrayList<Cursor>>() {}));
+                    } catch (IOException e) {
+                        writeProblemResponse(response, outputStream, HttpStatus.BAD_REQUEST.value(),
+                                new Problem("incorrect syntax of X-nakadi-cursors header"));
+                        return;
+                    }
+                }
+
+                // check that offsets are not out of bounds
+                if (cursors.isPresent() && !topicRepository.areCursorsCorrect(topic, cursors.get())) {
+                    writeProblemResponse(response, outputStream, HttpStatus.UNPROCESSABLE_ENTITY.value(),
+                            new Problem("cursors are not valid"));
+                    return;
+                }
+
+                // convert cursors to map; if no cursors provided - read from the newest available events
+                final Map<String, String> streamCursors = cursors
+                        .orElseGet(() -> topicRepository
+                                .listPartitions(topic)
+                                .stream()
+                                .map(pInfo -> new Cursor(pInfo.getPartitionId(), pInfo.getNewestAvailableOffset()))
+                                .collect(Collectors.toList()))
+                        .stream()
+                        .collect(Collectors.toMap(
+                                Cursor::getPartition,
+                                Cursor::getOffset));
+
+                final EventStreamConfig streamConfig = EventStreamConfig
+                        .builder()
+                        .withTopic(topic)
+                        .withCursors(streamCursors)
+                        .withBatchLimit(batchLimit)
+                        .withStreamLimit(streamLimit)
+                        .withBatchTimeout(batchTimeout)
+                        .withStreamTimeout(streamTimeout)
+                        .withBatchKeepAliveLimit(batchKeepAliveLimit)
+                        .build();
+
+                response.setStatus(HttpStatus.OK.value());
+
+                final String acceptEncoding = request.getHeader("Accept-Encoding");
+                final boolean gzipEnabled = acceptEncoding != null && acceptEncoding.contains("gzip");
+                final OutputStream output = gzipEnabled ? new FlushableGZIPOutputStream(outputStream) : outputStream;
```

Comment:
ok, this way of handling gzip is very messy: 
- supporting gzip controls the whole data transport and should not have to be dealt with explicitly this deep in the streaming). ARUHA-84 created for this.
- we end up passing this stream around and having to close optionally. this is made worse by the fact that it's being written to the stream in several different places.